packages/cms-cli/commands/logs.js

```diff
@@ -13,27 +14,78 @@ const { logDebugInfo } = require('../lib/debugInfo');
 const { loadConfig, checkAndWarnGitInclusion } = require('@hubspot/cms-lib');
 const { logger } = require('@hubspot/cms-lib/logger');
 const { logApiErrorInstance } = require('@hubspot/cms-lib/errorHandlers');
-const { toFile, outputLogs } = require('@hubspot/cms-lib/lib/logs');
+const { outputLogs } = require('@hubspot/cms-lib/lib/logs');
 const { getFunctionByPath } = require('@hubspot/cms-lib/api/function');
 const {
   getFunctionLogs,
   getLatestFunctionLog,
 } = require('@hubspot/cms-lib/api/results');
+const { base64EncodeString } = require('@hubspot/cms-lib/lib/encoding');
 
 const COMMAND_NAME = 'logs';
+const TAIL_DELAY = 5000;
 
-function getLogs(program) {
+const makeSpinner = (functionPath, portalIdentifier) => {
+  return ora(
+    `Tailing logs for '${functionPath}' on portal '${portalIdentifier}'.\n`
+  );
+};
+
+const makeTailCall = (portalId, functionId) => {
+  return async (after, spinner) => {
+    const latestLog = await getFunctionLogs(portalId, functionId, { after });
+    if (latestLog.results.length) {
+      spinner.clear();
+      outputLogs(latestLog);
+    }
+
+    return latestLog.paging.next.after;
+  };
+};
+
+const tailLogs = async (portalId, functionId, functionPath, portalName) => {
+  const tailCall = makeTailCall(portalId, functionId);
+  const spinner = makeSpinner(functionPath, portalName || portalId);
+  let after;
+
+  spinner.start();
+
+  try {
+    const latestLog = await getLatestFunctionLog(portalId, functionId);
+    after = base64EncodeString(latestLog.id);
+  } catch (e) {
+    // A 404 means no latest log exists(never executed)
+    if (e.statusCode !== 404) {
+      logApiErrorInstance(e, {
+        functionPath,
+        portalId,
+      });
+    }
+  }
+
+  return new Promise(() => {
```

Comment:
Something doesn't look quite right when it comes to the return value of a Promise here and what the calls are doing below with the return value. In general, we should also favor `async` over `new Promise()` unless there is a particular reason.

Reply:
I agree, this is unusual. Basically, the thread is being blocked, waiting for a promise that never resolves. This is done so we can tail the logs indefinitely, until the process is manually killed.

We could [spawn a child process](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) that would start the tailing, which would achieve the same result in a less hacky way.

Thoughts @gcorne @corymartin ?

Reply:
This is working POC code.

```javascript
async function fetchFunctionLogs() {
  return Promise.resolve((Math.random() * 1000) | 0);
}

const INTERVAL = 2000;

async function delayFetch() {
  return new Promise(resolve => {
    setTimeout(() => {
      debugger;
      fetchFunctionLogs().then(resolve);
    }, INTERVAL);
  });
}

async function* checkLogs() {
  for (;;) {
    const result = await delayFetch();
    yield result;
  }
}

async function tail() {
  for await (const log of checkLogs()) {
    console.log('>>>', log);
  }
}

tail();
```

Reply:
Try running that in Node env or in Chrome console.

Reply:
@miketalley One way to handle it is to not return from the async function in the `logs` subcommand action. I think one way of thinking that might lead to a solution is to consider how to handle error case. What happens if the user doesn't have permission to fetch logs or there is a network interruption while tailing?

Reply:
This has been improved greatly.  I removed the unnecessary promise.

Looking at the generator a bit more and comparing it with the current code, I think it creates unnecessary complication without a major benefit. Take a look at the new code and LMK if you don't agree @corymartin.